File: internal/collector/collector.go
package collector

import (
	"context"
	"log"
	"time"

	"github.com/Ruscigno/stock-screener/internal/config"
	"github.com/Ruscigno/stock-screener/internal/datasource/yahoo"
	"github.com/Ruscigno/stock-screener/internal/storage"
	"github.com/Ruscigno/stock-screener/internal/timeframe"
)

type Collector struct {
	store storage.Store
	src   *yahoo.Client
	cfg   *config.Config
}

func New(store storage.Store, src *yahoo.Client, cfg *config.Config) *Collector {
	return &Collector{store: store, src: src, cfg: cfg}
}

// nativeTimeframes reduces the configured timeframes to the distinct native
// ones that must be fetched (derived TFs are computed at query time, but their
// parent native TF must be collected).
func nativeTimeframes(tfs []string) []string {
	seen := map[string]bool{}
	var out []string
	add := func(name string) {
		if !seen[name] {
			seen[name] = true
			out = append(out, name)
		}
	}
	for _, name := range tfs {
		tf, ok := timeframe.Get(name)
		if !ok {
			continue
		}
		if tf.Native {
			add(tf.Name)
		} else {
			add(tf.Parent)
		}
	}
	return out
}

// dropUnclosed removes trailing candles whose bar end (Time + barDur) is after
// now (i.e. still forming).
func dropUnclosed(candles []yahoo.Candle, barDur time.Duration, now time.Time) []yahoo.Candle {
	out := candles
	for len(out) > 0 {
		last := out[len(out)-1]
		if last.Time.Add(barDur).After(now) {
			out = out[:len(out)-1]
		} else {
			break
		}
	}
	return out
}

// CollectOnce fetches every (symbol, native TF) once and upserts the bars.
// Per-item errors are logged and collected, not fatal.
func (c *Collector) CollectOnce(ctx context.Context) []error {
	var errs []error
	natives := nativeTimeframes(c.cfg.Timeframes)
	now := time.Now()
	for _, symbol := range c.cfg.Stocks {
		for _, tfName := range natives {
			tf, _ := timeframe.Get(tfName)
			from, _, err := c.store.LastBarTime(ctx, symbol, tfName)
			if err != nil {
				errs = append(errs, err)
				continue
			}
			candles, err := c.src.Fetch(ctx, symbol, tf.YahooInterval, from)
			if err != nil {
				log.Printf("collect %s %s: %v", symbol, tfName, err)
				errs = append(errs, err)
				continue
			}
			if c.cfg.Collector.UseClosedBarsOnly {
				candles = dropUnclosed(candles, tf.BarDuration, now)
			}
			bars := make([]storage.Bar, 0, len(candles))
			for _, cd := range candles {
				bars = append(bars, storage.Bar{
					Symbol: symbol, Timeframe: tfName, Time: cd.Time,
					Open: cd.Open, High: cd.High, Low: cd.Low, Close: cd.Close, Volume: cd.Volume,
				})
			}
			if err := c.store.UpsertBars(ctx, bars); err != nil {
				errs = append(errs, err)
				continue
			}
			log.Printf("collected %s %s: %d bars", symbol, tfName, len(bars))
			time.Sleep(200 * time.Millisecond) // gentle pacing for Yahoo
		}
	}
	return errs
}

// Run collects on a ticker until ctx is cancelled. Uses the intraday refresh
// cadence as the loop tick (the smaller of the two).
func (c *Collector) Run(ctx context.Context) {
	tick := time.Duration(c.cfg.Collector.Refresh.Intraday)
	if tick <= 0 {
		tick = 15 * time.Minute
	}
	c.CollectOnce(ctx)
	t := time.NewTicker(tick)
	defer t.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-t.C:
			c.CollectOnce(ctx)
		}
	}
}

File: stock_screener/collector.py
from __future__ import annotations
import logging
import time
from stock_screener import timeframes, datasource
from stock_screener.config import Config

log = logging.getLogger("stock_screener.collector")

def native_timeframes(tfs: list[str]) -> list[str]:
    seen, out = set(), []
    for name in tfs:
        tf = timeframes.get(name)
        if tf is None:
            continue
        n = tf.name if tf.native else tf.parent
        if n not in seen:
            seen.add(n)
            out.append(n)
    return out

class Collector:
    def __init__(self, store, cfg: Config, fetch=datasource.fetch):
        self.store = store
        self.cfg = cfg
        self._fetch = fetch

    def collect_timeframes(self, tf_names: list[str]) -> list[Exception]:
        errs: list[Exception] = []
        for symbol in self.cfg.stocks:
            for tf_name in tf_names:
                tf = timeframes.get(tf_name)
                try:
                    start = self.store.last_bar_time(symbol, tf_name)
                    df = self._fetch(symbol, tf.yahoo_interval, tf.bar_seconds, start,
                                     self.cfg.collector.use_closed_bars_only)
                    self.store.upsert_bars(symbol, tf_name, df)
                    log.info("collected %s %s: %d bars", symbol, tf_name, len(df))
                    time.sleep(0.2)
                except Exception as e:  # noqa: BLE001 - per-item isolation
                    log.error("collect %s %s: %s", symbol, tf_name, e)
                    errs.append(e)
        return errs

    def collect_once(self) -> list[Exception]:
        return self.collect_timeframes(native_timeframes(self.cfg.timeframes))
